Scripting: Fix script docs not being searchable without manually recompiling scripts

anvilfolk opened the pull request:
This PR adds a script documentation cache in the project folder. It is loaded at alongside native documentation caches. This makes scripts fully accessible through Search Help, including their members, etc, right from project start, without having to compile every single script manually to access its docs.

Testing with GDScript but, in theory, should work for any scripting language that has documentation support. Might be worth splitting the cache into a per-language basis to reduce conflicts, though if there are conflicts with e.g. a MyClass in GDScript and MyClass in C#, that may already lead to problems with the current documentation system.

~~Ensuring file deletions when Godot closed trigger EditorFileSystem's documentation updates is done by #95965, until then the cache will persist docs from deleted scripts~~ Done.

Diagram of call behavior that I needed to try to organize this in my brain:

![image](https://github.com/user-attachments/assets/c5770c79-ba59-44af-9093-3c4bce46ca94)

Grey background runs on whatever thread called it. Orange background runs in `worker_thread`. Green background runs on `loader_thread`, and can include `ResourceLoader::load()` calls, orange cannot. Blue background represents waiting for one shot `sources_updated` signals depending on whether `EditorFileSystem::is_scanning()` is true.

Implementation details:
- Doc tasks that don't use `ResourceLoader::load()` to generate docs always happens in `EditorHelp::worker_thread`.
- Regenerating script docs using `ResourceLoader::load()` happen in `EditorHelp::loader_thread`. _This prevents deadlocks where the main thread needs doc info and waits for `worker_thread`, but `worker_thread` is waiting for main thread to dispatch loading tasks._
- Script cache should not reload when GDExtension docs change, overwriting doc changes of the current session ❗ someone with GDExtensions please test!❗ .
- If cache isn't present, `worker_thread` connects to `EditorFileSystem::filesystem_changed` signal, indicating the end of `EditorFileSystem`'s scan, and ends its execution. It spools back up once the signal is fired to regenerate the cache, this time using `EditorFileSystemDirectory`, therefore not accessing underlying OS filesystem calls.
- Cache is saved & loaded as-is in memory or in the disk. We let `EditorFileSystem` catch changes like deleted/added files to ensure docs are kept up to date.
- Cache only saves on exit, since it requires bulk-saving every script doc to disk as a Resource. That feels like too much to do every time a script is saved/compiled to keep disk & memory docs consistent.
- Added `DocTool` method forwarding to `EditorHelp` to help maintain cache consistency. It is meant to track whether changes affect script docs or other docs. These forwarded methods should be used from now on instead of e.g. `get_doc_data()->add_doc()`.
- Disk cache is deleted after successful load. It saves on editor exit, if successfully loaded/regenerated previously.
- Somewhat unrelated but removed quotes around documentation pages in the script list, e.g. `"new_script.gd"` becomes `new_script.gd`, since this resulted in messy truncations of paths (to test, create `new_folder/new_script.gd` and `new_script.gd` and open both their documentations on stable.

Fixes #72406, fixes #86577, fixes #72952 (for GDScript, C# docs are not supported yet).

Added coauthorship with @Hilderin because of their wonderful and insightful help in designing and testing this <3

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

anvilfolk committed:
Scripting: Add script documentation cache to project

This PR adds a script documentation cache in the project folder.
It is loaded at alongside native documentation caches. This makes
scripts fully accessible through Search Help, including their
members, etc, right from project start, without having to compile
every single script.

Co-authored-by: Hilderin <81109165+Hilderin@users.noreply.github.com>

anvilfolk commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -2232,6 +2202,36 @@ void ScriptEditor::_update_script_names() {
 		}
 	}
 
+	Vector<String> disambiguated_script_names;
+	Vector<String> full_script_paths;
+	for (int j = 0; j < sedata.size(); j++) {
+		String name = sedata[j].name.replace("(*)", "");
+		ScriptListName script_display = (ScriptListName)(int)EDITOR_GET("text_editor/script_list/list_script_names_as");
+		switch (script_display) {
+			case DISPLAY_NAME: {
+				name = name.get_file();
+			} break;
+			case DISPLAY_DIR_AND_NAME: {
+				name = name.get_base_dir().get_file().path_join(name.get_file());
+			} break;
+			default:
+				break;
+		}
+
+		disambiguated_script_names.append(name);
+		full_script_paths.append(sedata[j].tooltip);
+	}
+
+	EditorNode::disambiguate_filenames(full_script_paths, disambiguated_script_names);
+
+	for (int j = 0; j < sedata.size(); j++) {
+		if (sedata[j].name.ends_with("(*)")) {
+			sedata.write[j].name = disambiguated_script_names[j] + "(*)";
+		} else {
+			sedata.write[j].name = disambiguated_script_names[j];
+		}
+	}
+
```
I'm not sure why the doc/script name disambiguation code was inside the `for` loop which collects elements for the script/doc list. If the last tab was a documentation page, the disambiguation code would not run at all, resulting in weird artifacts.

I think running name/path disambiguation code makes more sense after scripts & docs have been collected, but didn't dig too deeply to see if it only works when disambiguation is called every loop. Didn't notice any problems in my admittedly relatively minor testing.

anvilfolk commented on editor/editor_help.cpp:
```diff
@@ -2935,7 +2948,64 @@ void EditorHelp::_gen_extensions_docs() {
+		return;
+	}
+
+	// Force compile all scripts, which should generate their documentation.
+	WARN_PRINT_ED("Script documentation cache not found. Regenerating it may take a while.");
+	HashMap<String, bool> is_extension_script; // Caches ScriptServer results so it's not being queried for every file, only for every extension.
+	List<Ref<DirAccess>> dirs;
+	dirs.push_front(DirAccess::create(DirAccess::ACCESS_RESOURCES));
+	while (!dirs.is_empty()) {
+		Ref<DirAccess> dir = dirs.front()->get();
+		dirs.pop_front();
+		for (const String file : dir->get_files()) {
+			if (file.begins_with("."))
+				continue;
+			if (!is_extension_script.has(file.get_extension()))
+				is_extension_script[file.get_extension()] = ScriptServer::get_language_for_extension(file.get_extension()) != nullptr;
+
+			if (is_extension_script[file.get_extension()]) {
+				Ref<Script> scr = ResourceLoader::load(file);
+				if (scr.is_valid())
+					for (DocData::ClassDoc cd : scr->get_documentation())
+						doc->add_doc(cd);
+			}
+		}
+
+		for (const String subdir : dir->get_directories()) {
+			if (!subdir.begins_with("."))
+				dirs.push_front(DirAccess::open(DirAccess::get_full_path(dir->get_current_dir() + "/" + subdir, DirAccess::ACCESS_RESOURCES)));
+		}
+	}
+
+	save_script_doc_cache();
+}
```
This goes through every single file in `res://` and forces it to recompile so we can get its documentation. I think this is the most efficient I can make it if the cache is not present. But I'd love more eyes on this to see if I'm missing something.

Alternatively, this can be done somewhere in `EditorFileSystem`.

anvilfolk replied:
For posterity, I did try moving this code into `EditorFileSystem`, but we run into threading issues. If we call `EditorFileSystem` and use existing methods like `_update_script_documentation()`, then we cannot be running from `EditorHelp::worker_thread`, which we are and want to.

I don't think it's worth the threading hassle, we can just have walk through the filesystem instead of using the project's file cache :)

anvilfolk replied:
This has been changed to waiting until `EditorFileSystem` is done with its scan and using its datastructures.

dalexeev commented:
> * Somewhat unrelated but removed quotes around documentation pages in the script list, e.g. `"new_script.gd"` becomes `new_script.gd`, since this resulted in messy truncations of paths (to test, create `new_folder/new_script.gd` and `new_script.gd` and open both their documentations on stable.

I think it's better to fix this by checking the tab type. Doc tabs should never truncate the name, only script tabs.

anvilfolk commented:
> > * Somewhat unrelated but removed quotes around documentation pages in the script list, e.g. `"new_script.gd"` becomes `new_script.gd`, since this resulted in messy truncations of paths (to test, create `new_folder/new_script.gd` and `new_script.gd` and open both their documentations on stable.
> 
> I think it's better to fix this by checking the tab type. Doc tabs should never truncate the name, only script tabs.

Why the difference? You can still get the full path by hovering and it's at the top of the help document. This way it feels like behavior is at least consistent?

anvilfolk commented on editor/editor_help.cpp:
```diff
@@ -2935,7 +2948,65 @@ void EditorHelp::_gen_extensions_docs() {
+
+	// Force compile all scripts, which should generate their documentation.
+	WARN_PRINT("Script documentation cache not found. Regenerating it may take a while.");
+	HashMap<String, bool> is_extension_script; // Caches ScriptServer results so it's not being queried for every file, only for every extension.
```
This probably doesn't make performance that much better. I wonder whether it's worth the extra complication.

Hilderin commented on editor/editor_help.cpp:
```diff
@@ -2935,7 +2948,65 @@ void EditorHelp::_gen_extensions_docs() {
+			for (int i = 0; i < classes.size(); i++)
+				doc->add_doc(DocData::ClassDoc::from_dict(classes[i]));
+		}
+		return;
```
Maybe this return should be into the 'if (script_doc_cache_res.is_valid()) {'. If the file is corrupt or invalid, it then should regenerate the cache?

anvilfolk replied:
Yes, great catch! <3

Hilderin commented on editor/editor_help.cpp:
```diff
@@ -2935,7 +2948,65 @@ void EditorHelp::_gen_extensions_docs() {
+
+		for (const String &subdir : dir->get_directories()) {
+			if (!subdir.begins_with("."))
+				dirs.push_front(DirAccess::open(DirAccess::get_full_path(dir->get_current_dir() + "/" + subdir, DirAccess::ACCESS_RESOURCES)));
```
`dir->get_current_dir().path_join(subdir)` is preferable then using ` + "/" + `, that way it will concatene with "\" on Windows.

anvilfolk replied:
Thank you! I remembered there was something like that then forgot to follow up!

Hilderin commented on editor/editor_help.cpp:
```diff
@@ -2935,7 +2948,65 @@ void EditorHelp::_gen_extensions_docs() {
+				is_extension_script[file.get_extension()] = ScriptServer::get_language_for_extension(file.get_extension()) != nullptr;
+
+			if (is_extension_script[file.get_extension()]) {
+				Ref<Script> scr = ResourceLoader::load(dir->get_current_dir() + "/" + file);
```
`dir->get_current_dir().path_join(file)` is preferable then using ` + "/" + `, that way it will concatene with "\" on Windows.

Hilderin commented on editor/editor_help.cpp:
```diff
@@ -2935,7 +2948,65 @@ void EditorHelp::_gen_extensions_docs() {
+	}
+
+	// Force compile all scripts, which should generate their documentation.
+	WARN_PRINT("Script documentation cache not found. Regenerating it may take a while.");
```
I'm not a sure about the warning, it's usually a sign that something went wrong. Maybe a standard print saying that while the cache is regenerating the help on script will not be available and when the help is finished regenerating, printing another message saying that the documentation is ready?

anvilfolk replied:
Yes, you're right, I will use `print_verbose`, which is consistent with when there is no native help cache!

Hilderin commented:
Currently, if the get_script_doc_cache_full_path file does not exists, the _load_script_doc_cache method will parse all the files on the disk to find all the script. That will happen at the same time as the file scanning in EditorFileSystem. We received complains that loading a project with thousands of file is currently slow. With this double scanning at the same time, it will not be helping. Also, some logic to scan the disk is not implemented here, see `EditorFileSystem::_should_skip_directory`.

What I suggest is that if the cache is not there, connect a callback on the "filesystem_changed" signal from `EditorFileSystem` singleton and when it's emitted, start a thread to load the documentation. Another way is no wait until `EditorFileSystem::get_singleton()->is_scanning()` is false.

That way you also will be able to get access to all the project files in `EditorFileSystem` like the `FileSystemDock` does and as a bonus, the type of the file will tell you if it's a script using `ClassDB::is_parent_class(type, "Script")`.

anvilfolk commented:
I see the issue, thanks for explaining! I'm worried that a one-off connection to `filesystem_changed` may catch events that do not need a full cache regeneration, like deleting/moving a file while Godot is closed. That should be handled by individual file actions, which do not cause full script doc regen.

I might end up detaching script doc regeneration from the current doc benchmarks, and probably just have the thread wait for scanning to finish, then avoid using `DirAccess` which does a bunch of underlying OS calls, and instead use `EditorFileSystem`'s data like you suggested.

anvilfolk commented:
I'll test this further tomorrow, but addressed @Hilderin's concerns by using their solution of using a one-off connection to `filesystem_changed` if a scan is running, and respooling up the worker thread to go through `EditorFileSystemDirectory` and reload scripts once the scan is done.

Would appreciate a review for folks more attuned to multithreading issues :)

anvilfolk commented:
Ran this and got `Caller thread can't call this function in this node (%s). Use call_deferred() or call_thread_group() instead.` pointing to `EditorFileSystem` a single time after recompiling but haven't been able to replicate it. Again, multithreading expert review wanted 😅 

Behavior right now is:
- `EditorHelp::regenerate_script_doc_cache()`  is called from main thread since if called from `worker_thread` it might wait on itself
- `EditorHelp::regenerate_script_doc_cache()` checks if scan is ongoing, if so connects itself to`EditorFileSystem::filesystem_changed` then returns. It will start generating the cache in the `worker_thread`, running `_gen_script_doc_thread` when it is finally called with no ongoing scan.
- `_gen_script_doc_thread` calls `EditorFileSystem::get_singleton()->get_filesystem()` and `ResourceLoader::load()` but that's about it.

Hilderin commented on editor/editor_node.cpp:
```diff
@@ -813,6 +814,11 @@ void EditorNode::_notification(int p_what) {
+
+		case NOTIFICATION_PREDELETE: {
+			// Cannot be done in destructor since saving a resource emits EditorNode signals using memory that's being destroyed.
+			EditorHelp::save_script_doc_cache();
```
I tried with a project with 3000 script files and it took a while to load the project. I was able to close the editor before all the script were parsed. Maybe it could be interesting to have an indicator that the script cache is fully loaded and check that indicator before saving the cache. Maybe even removing the cache file if not completely loaded when quitting the editor? I'm also wondering if it should be interesting to remove the cache file when a new script is modified preventing the cache to be outdated if the editor crash or something.

anvilfolk replied:
Some thoughts!
- Addressing the first concern: I can make loading and saving use the worker thread, so they can't happen simultaneously. Or at least, I can make the save on Godot exit wait for the worker thread.
- If we delete the cache as soon as a script is modified, we force-recompile all scripts on basically every run of the editor since most folks edit a script at one point or another, so that doesn't feel viable.
- With your PR, we should be able to keep cache consistent by getting an update whenever a a file changes, so the docs are re-added, then saved on exit.
- I like the idea of deleting the cache file just before saving it, it safeguards against errors on editor shutdown.

Hilderin replied:
> Addressing the first concern: I can make loading and saving use the worker thread, so they can't happen simultaneously. Or at least, I can make the save on Godot exit wait for the worker thread.

I think you can keep it very simple by adding a global bool at false by default and to set it at true when the documentation for all scripts has been loaded or the cache is loaded. You skip writing the cache if the bool is not true on editor exit. That way, the cache will never contain partial documentation. I strongly suggest you don't use thread on exit, the thread will never finish.

> If we delete the cache as soon as a script is modified, we force-recompile all scripts on basically every run of the editor since most folks edit a script at one point or another, so that doesn't feel viable.

Not really because it's resaved on editor exit. The only time the documentation should be completely regenerated is if the editor crash. I think it's better to generate the doc then having a partial documentation. 

Personally I would try something like that:
- Global bool doc_scripts_ready = false
- Load cache or regenerate script doc
- Set doc_scripts_ready = true when done
- When new doc arrives, delete the cache on disk if exists (this cache is deprecated and should never be reused)
- On exit if doc_scripts_ready = true, write the cache on disk

That way:
- The cache on disk is never deprecated, worst case then doc needs to be regenerated on next startup
- It's simple, easy to implement and debug
- Manage crashes
- Manage exit before all the project is loaded
- When there's no cache you shoud always have a cache on disk that is up to date on the next startup

anvilfolk replied:
Done!

Hilderin commented on editor/editor_help.cpp:
```diff
@@ -2935,7 +2952,84 @@ void EditorHelp::_gen_extensions_docs() {
+		if (ClassDB::is_parent_class(dir->get_file_type(i), SNAME("Script"))) {
+			Ref<Script> scr = ResourceLoader::load(dir->get_file_path(i));
+			if (scr.is_valid()) {
+				for (DocData::ClassDoc cd : scr->get_documentation()) {
```
Always use iterators when available:
```for (const DocData::ClassDoc &cd : scr->get_documentation()) {```

Hilderin commented on editor/editor_help.cpp:
```diff
@@ -2935,7 +2952,84 @@ void EditorHelp::_gen_extensions_docs() {
+		Ref<Resource> script_doc_cache_res = ResourceLoader::load(get_script_doc_cache_full_path());
+		if (script_doc_cache_res.is_valid()) {
+			Array classes = script_doc_cache_res->get_meta("classes", Array());
+			for (int i = 0; i < classes.size(); i++) {
+				doc->add_doc(DocData::ClassDoc::from_dict(classes[i]));
+			}
```
```cpp
for (const Dictionary &dict : classes) {
    doc->add_doc(DocData::ClassDoc::from_dict(dict));
}
```

Hilderin reviewed (changes requested):
That seems to work very well, even on large project!! The loading and saving of the cache is quick and seems efficient. I did not encounter any thread problem on my side.

anvilfolk commented:
Great review, thank you! Addressed most of the feedback and pushed.

I want to be more clear-headed when thinking through the final issue you mentioned!

anvilfolk commented:
Added some forwarding methods to `EditorHelp` from `DocTools`, which are there to ensure the script cache remains consistent. Whenever in-memory script docs change, the disk cache is deleted to ensure consistency. Also added the boolean to check whether cache has been successfully loaded and can be written to disk!

Hilderin commented on editor/editor_help.cpp:
```diff
@@ -2859,6 +2861,35 @@ String EditorHelp::get_cache_full_path() {
+
+void EditorHelp::remove_doc(const String &p_class_name) {
+	DocTools *dt = get_doc_data();
+	if (dt->has_doc(p_class_name) && dt->class_list[p_class_name].is_script_doc) {
+		_delete_script_cache();
+	}
+	dt->remove_doc(p_class_name);
```
```suggestion
	if (dt->has_doc(p_class_name)) {
		if(dt->class_list[p_class_name].is_script_doc) {
			_delete_script_cache();
		}
		dt->remove_doc(p_class_name);
	}
```

anvilfolk replied:
Your point is taken, but I'm not sure about the suggestion! We only want to delete the script doc cache when the doc being removed is from a script though, right? The `EditorHelp::has/add/remove_doc()` interface is now meant to be used instead of `EditorHelp::get_doc_data()->has/add/remove_doc()`, and it's responsible for keeping cache consistency.

Either way, I moved some things around here.

Hilderin commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -3611,8 +3611,8 @@ void ScriptEditor::clear_docs_from_script(const Ref<Script> &p_script) {
 		const DocData::ClassDoc &doc = documentations.get(j);
-		if (EditorHelp::get_doc_data()->has_doc(doc.name)) {
-			EditorHelp::get_doc_data()->remove_doc(doc.name);
+		if (EditorHelp::has_doc(doc.name)) {
```
Not need to call `EditorHelp::has_doc` because it's done in `EditorHelp::remove_doc`.

Hilderin commented on editor/editor_help.cpp:
```diff
@@ -2935,7 +2976,95 @@ void EditorHelp::_gen_extensions_docs() {
+}
+
+void EditorHelp::_delete_script_cache() {
+	if (FileAccess::exists(get_script_doc_cache_full_path()))
+		DirAccess::remove_file_or_error(ProjectSettings::get_singleton()->globalize_path(get_script_doc_cache_full_path()));
```
```suggestion
	if (FileAccess::exists(get_script_doc_cache_full_path())) {
		DirAccess::remove_file_or_error(ProjectSettings::get_singleton()->globalize_path(get_script_doc_cache_full_path()));
	}
```

Hilderin commented:
Works great!! Good job!!

I tested this PR merging locally #95965 and #96002

There's a last problem where the doc loading thread can be reloading the doc script cache while the EditorFileSystem is scanning and updating the doc. That could lead to problems where the cache is loaded after the remove_doc or add_doc was called so the last version of the script is still in the documentation. Worst, we could try to read the cache while trying to delete it and vice versa. I suggest the script cache loading should be postponed after the first scan. You already done the logic when the cache is not present to wait for signal `filesystem_changed` if is_scanning is true.

Tested:
- Modifying scripts in the builtin Script Editor
- Modifying scripts in a external editor connected to LSP server (ex: VSCode)
- Modifying scripts in a external editor not connected to LSP server (ex: NotePad) (Was not working without #96002. I also found a bug there that I just fixed in the #96002)
- Modifying scripts in a external editor while the doc is opened in Godot (automatically updates with returning to godot) 
- Modifying scripts while the editor is closed: Not working perfectly, see comment above.
- Removing scripts while the editor is closed: Not working perfectly, see comment above.

Hilderin commented:
Good news, this prerequisite PR has been merged: #95965

anvilfolk commented:
> Good news, this prerequisite PR has been merged: #95965

Perfect, rebased, fixed a couple of things and pushed! Been too sleepy/tired to work on this, but I don't know that we have an immediately obvious solution to the problem you mentioned.

If we want to wait to load the cache until the first scan is complete, maybe `EditorFileSystem_update_script_documentation()` has already called `EditorHelp::add/remove_doc`, and those changes will be overwritten when the cache is read. I believe the dependencies we're looking at are:

If regenerating the cache: 1) First scan -> 2) Regenerate cache from scratch
If loading existing cache: 1) Load cache -> 2) EditorFileSystem updates script docs

I'm worried about creating a race condition where EditorHelp is waiting for first scan, but before first scan finishes, it waits for EditorHelp to load the cache. Will try to give this some more thought when life allows!

SanchotEm commented:
This entire summer i couldn't understand what was the issue with custom Documentations and why i had to resav scripts in order to make them working again.
I'm really hoping your fix will be implemented asap in next updates! Tysm for contibuting into Godot better future!

anvilfolk commented:
Been working on this and running into a couple of problems. I did rebase onto `master` today, so I suspect it's from recently introduced changes elsewhere. Maybe from the progress bar or something about loading many files. I don't think this new version of the PR is significantly different.

If the cache has not been generated, then Godot freezes then runs out of memory & crashes here with a project with 4096 .gd files:
![image](https://github.com/user-attachments/assets/0fad0d08-3339-4d99-b260-f4c8278c2ffe)
![image](https://github.com/user-attachments/assets/98127d76-1053-4903-9a1e-c1fb1d9fb06f)

If the cache has been generated and is just loaded, no such issue occurs.



Also, if a script is updated using an external editor while Godot is open, the script file updates but not the open doc screen. This is true both with the previous (non-rebased & not updated) version of the as well as the new one. Can't tell why. I seem to remember it working. `EditorHelp::add_doc(docs[j]);` in `EditorFileSystem::_update_script_documentation()` seems to be called _before_ the popup asking to ignore local changes and reload.

I'll keep investigating.

anvilfolk commented:
Rebased once more and the crash seems to have been fixed, even in a project wit 8192 files. Will dig into the script updating but not the script doc!

Hilderin commented on editor/editor_help.cpp:
```diff
@@ -686,7 +668,7 @@ void EditorHelp::_pop_code_font() {
 
 Error EditorHelp::_goto_desc(const String &p_class) {
 	// If class doesn't have docs listed, attempt on-demand docgen
```
```suggestion
	// If class doesn't have docs listed.
```
Or completely remove the comment, it does not gives any useful information.

Hilderin commented on editor/editor_help.cpp:
```diff
@@ -2974,7 +3003,138 @@ void EditorHelp::_gen_extensions_docs() {
+	} else {
+		print_verbose("Script doc cache is corrupted. Regenerating it instead.");
+		_delete_script_doc_cache();
+		callable_mp_static(EditorHelp::regenerate_script_doc_cache).call_deferred();
```
```suggestion
		callable_mp_static(EditorHelp::regenerate_script_doc_cache).call_deferred();
		return;
```
I think there should be a `return` after the call deferred to prevent `script_docs_loaded` to be true while the cache is regenerating.

anvilfolk replied:
Oh, great catch, thank you!

Hilderin commented:
Very very nice modifications that you made on your last commit. I like the way you postponed the cache loading until the first scan is done and that you keep track of the changes while scanning and apply them after the cache is loaded!!!

> Also, if a script is updated using an external editor while Godot is open, the script editor updates but not the open doc screen. This is true both with the previous (non-rebased & not updated) version of the as well as the new one. I thought it worked before, but doesn't seem to for me.

The PR #96002 fixes the issue where the doc is not updated in memory, but there's still a issue if the doc is opened in the script editor, it's not updated, the doc needs to be closed and reopened. I'm not sure it's supposed to be fixed in this current PR, I'll look into it in #96002.

Tested:
- Modifying scripts in the built-in Script Editor
- Modifying scripts in a external editor connected to LSP server (ex: VSCode)
- Modifying scripts in a external editor not connected to LSP server (ex: NotePad) (Was not working without #96002, not working completely, see comment above)
- Modifying scripts in a external editor while the doc is opened in Godot
- Modifying scripts while the editor is closed
- Removing scripts while the editor is closed
- Tested with a small project (4-5 scripts)
- Tested with a medium project (1000 scripts)

anvilfolk commented:
I have been able to replicate the crash. It happens when 
1) the editor is run with `--verbose`, and
2) the cache is not present and therefore regenerated, force-loading all scripts.

It tries to print one line for each of 8192 resources being loaded onto the console, which I believe causes it to run out of memory.

anvilfolk commented:
Tested the following with a 8000+ .gd file project:
No filecache, no doc cache: works ✅ 
Filecache, no doc cache: works ✅ 
No filecache, doc cache: works, but does regenerate the cache which makes sense ✅ 
Filecache, doc cache: works ✅ 

Godot open:
- Change file in notepad, then discard changes: works ✅ 
- Change file in notepad, then accept changes: works ✅ but the .gd file itself is not updated though the documentation is, bisected to #96002
- Delete file: works ✅ 
- Add new file: works ✅ 

Godot closed:
- Change file in notepad: occasionally does not update documentation ⚠️ 
- Delete file: works
- Add new file: works

I can no longer replicate the issue above, though it happened consistently for a few minutes. Tried with many combinations of different .gd files & docs open or closed in the ScriptEditor.

I thought it might be from simultaneous reads/writes to `_docs_to_add`, but I don't think that's possible since `load_script_doc_cache` always waits for the `filesystem_changed` signal. Maybe if `load_script_doc_cache` is called after the filesystem scan is done but before `EditorFileSystem::_update_script_documentation()` is called, since there is a single situation in which the signal is emitted a while later, using `call_deferred(SNAME("emit_signal"), "filesystem_changed"); // Update later` in `EditorFileSystem::update_files`?

That said, I'm not sure it's worth adding mutexes and locking for a low probability situation which can be fixed manually relatively easily?

Hilderin commented:
> Change file in notepad: occasionally does not update documentation 

I think I created some regressions with #96002, it could be caused by that. This PR could maybe fix the intermittent problem. I'm not sure it should be investigated further here. It's being addressed in #97168

> That said, I'm not sure it's worth adding mutexes and locking for a low probability situation which can be fixed manually relatively easily?

I'm pretty sure it's not necessary to use mutex because the `EditorFileSystem` and the `ScriptEditor` reload resources and scripts in the main thread. Also, the situation you are describing should not happen to often. Worse case scenario, the user resave the script.

Hilderin commented on editor/editor_help.cpp:
```diff
@@ -2974,7 +3002,139 @@ void EditorHelp::_gen_extensions_docs() {
+}
+
+// Local function to avoid adding editor_file_system.h dependency to editor_help.h
+static void _reload_scripts_documentation(EditorFileSystemDirectory *dir, DocTools *doc) {
```
```suggestion
static void _reload_scripts_documentation(EditorFileSystemDirectory *p_dir, DocTools *p_doc) {
```

anvilfolk replied:
Will update the parameter names, good catch, missed that here! Remembered we can forward-declare classes so we don't need to add `editor_file_system.h` dependency to `editor_help.h` :)

anvilfolk commented:
> > Change file in notepad: occasionally does not update documentation
> 
> I think I created some regressions with #96002, it could be caused by that. This PR could maybe fix the intermittent problem. I'm not sure it should be investigated further here. It's being addressed in #97168 
> > That said, I'm not sure it's worth adding mutexes and locking for a low probability situation which can be fixed manually relatively easily?
> 
> I'm pretty sure it's not necessary to use mutex because the `EditorFileSystem` and the `ScriptEditor` reload resources and scripts in the main thread. Also, the situation you are describing should not happen to often. Worse case scenario, the user resave the script.

Updated PR with latest feedback!

And sounds great! I will try to get some time and look into those PRs, though `EditorFileSystem` is still sutble and confusing to me :)

I think you're right on this PR maybe not being the problem.

These are my attempts to understand the possible issues, just for documentation purposes. The issue I was thinking would happen if `EditorHelp::_load_script_doc_cache_thread` was running on worker thread while `EditorFileSystem::_update_script_documentation()` was running on other thread. This could mean some docs added through `EditorHelp::add_doc()` would be ignored/cleared without being added.

I followed whether `fileystem_changed` can be emitted before `_update_script_documentation()` is done running. It is only called in `_process_update_pending()`, which occurs in 3 places:
1. In `_update_scan_actions()`: if it returns true, always emitted after
2. In `update_files()`: always emitted after, but `_process_update_pending()` is only called when _not_ scanning.
3. In `reimport_files()`: always emitted after if it's scanning, and the cache always loads, so it's always emitted.

So we are sure `filesystem_changed` always emits _after_ `_update_script_documentation()` has run. But maybe the issue in 2), where `_process_update_pending()` is not called if `!is_scanning()`. `update_files()` is only called in `_update_scan_actions()`, which is handled in 1), or from `update_file()` (singular), which is called many places. Maybe `update_file()` is called somewhere outside of scanning, causing the docs to fail to update? I don't know.

Hilderin reviewed (approved):
Great work!! This PR will solve a lot of questions/issues with the script documentation not present after restarting the editor.

The author and me were was not able to reproduce to issue where the documentation was not updated in some situations. Maybe someone else while testing can reproduce.

dalexeev commented:
https://github.com/godotengine/godot/blob/3f2ef8d9c2d85c6dbe4de05a442eeca39e48346e/editor/editor_help.cpp#L3350

https://github.com/godotengine/godot/blob/3f2ef8d9c2d85c6dbe4de05a442eeca39e48346e/editor/editor_help.cpp#L3387

https://github.com/godotengine/godot/blob/3f2ef8d9c2d85c6dbe4de05a442eeca39e48346e/editor/editor_help.cpp#L3468

https://github.com/godotengine/godot/blob/3f2ef8d9c2d85c6dbe4de05a442eeca39e48346e/editor/editor_help.cpp#L3520

https://github.com/godotengine/godot/blob/3f2ef8d9c2d85c6dbe4de05a442eeca39e48346e/editor/editor_help.cpp#L3573

https://github.com/godotengine/godot/blob/3f2ef8d9c2d85c6dbe4de05a442eeca39e48346e/editor/editor_inspector.cpp#L3337

anvilfolk commented:
Copy-pasted from developer's chat since it might be relevant here.

@dalexeev, thanks for pointing out those comments on the gdscript docs PR! I'm not super certain what to do with it - I think things might be working as intended still. Scripts docs will not be cached (presumably since they can change so often) so will always get processed and returned by the functions that you highlighted.
We could try to start caching them and use some signal in order to keep the caches updated perhaps, but unless we find that is performance-critical it seems like more code for little gain. All of this shows up in `EditorHelpBit` - I'm not entirely sure what it is or where it's used, so it's hard for me to figure out performance details as well.

dalexeev commented:
@anvilfolk I just pointed out that there is another cache in `EditorHelpBit` and its behavior is now inconsistent with the `EditorHelp` cache. This may be confusing for future contributors, so we should look into whether it is possible to merge the caching between `EditorHelp` and `EditorHelpBit` to avoid extra entities and data duplication. Of course, this is material for a separate PR.

anvilfolk commented:
Yes, you're right. For documentation purposes, here's the caches I've noticed and my understanding of them.
- `EditorHelp` native doc cache: generated once and lives on disk in the user or %APPDATA% directory. Only changes with new Godot versions.
- `EditorHelp` script doc cache lives on disk in the project .godot folder to persist script documentation across runs of Godot. In memory it's just data under `DocTools`.
- `EditorHelpBit` seems to be a runtime cache, not existing on disk, not used for scripts, to avoid calling `DocTools` and doing some minor processing of its docs into `EditorHelpBit::HelpData`.

KoBeWi commented:
> Disk cache is deleted after successful load. It saves on editor exit, if successfully loaded/regenerated previously.

Why? When you don't modify it, it re-saves for no reason. You could keep the file and save only if it was modified.

Also I noticed folding for the cache is saved.
![image](https://github.com/user-attachments/assets/3182aa44-1538-4d80-8c43-081c594464ad)
Useless, but harmless. I wonder why it's saved though if I don't open the file.

AThousandShips commented on editor/editor_help.cpp:
```diff
@@ -2974,7 +3002,138 @@ void EditorHelp::_gen_extensions_docs() {
+	ERR_FAIL_COND_MSG(!ResourceLoader::exists(get_script_doc_cache_full_path()), "Error: cannot load script doc cache from inexistent file.");
+
+	Ref<Resource> script_doc_cache_res = ResourceLoader::load(get_script_doc_cache_full_path());
+	if (!script_doc_cache_res.is_valid()) {
```
```suggestion
	if (script_doc_cache_res.is_null()) {
```

AThousandShips commented on editor/editor_help.cpp:
```diff
@@ -34,11 +34,13 @@
 #include "core/core_constants.h"
 #include "core/extension/gdextension.h"
 #include "core/input/input.h"
+#include "core/io/dir_access.h"
```
```suggestion
```
Unnecessary

KoBeWi commented on editor/doc_tools.cpp:
```diff
@@ -369,6 +369,14 @@ void DocTools::remove_doc(const String &p_class_name) {
+void DocTools::remove_script_doc_by_path(const String &p_path) {
+	for (KeyValue<String, DocData::ClassDoc> &E : class_list) {
+		if (E.value.is_script_doc && E.value.script_path == p_path) {
+			remove_doc(E.key);
```
Shouldn't this `break`?

anvilfolk commented:
> > Disk cache is deleted after successful load. It saves on editor exit, if successfully loaded/regenerated previously.
> 
> Why? When you don't modify it, it re-saves for no reason. You could keep the file and save only if it was modified.

We don't have a way to change only part of the file, so the entire disk cache will be invalidated whenever any single script is saved. I made the call to simplify the code by avoiding return values, return parameters, and callbacks to `EditorHelp` from `DocTools` and `EditorHelp` functions. IMO it would be more code that's more prone to errors for the benefit of the usecase that someone never touches a script when working with Godot, as opposed to probably less than a second of extra time when exiting Godot.

> Also I noticed folding for the cache is saved. ![image](https://private-user-images.githubusercontent.com/2223172/369942387-3182aa44-1538-4d80-8c43-081c594464ad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxMDcyODAsIm5iZiI6MTcyNzEwNjk4MCwicGF0aCI6Ii8yMjIzMTcyLzM2OTk0MjM4Ny0zMTgyYWE0NC0xNTM4LTRkODAtOGM0My0wODFjNTk0NDY0YWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjNUMTU1NjIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTc4ZjNlODNjOTk3MzZiNjc0ZGFkYjBkNGQzN2FjZWJlMGJlMDgyMDY2YjJhYTAzNWM3MjVmOTI5OTNkNDg3NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.JILY8z6xxvgH7BpKfxuBkXP0HZ-TITt_Lp2sJggA8w8) Useless, but harmless. I wonder why it's saved though if I don't open the file.

Not sure what the folding is - it's my first time saving stuff to disk! Happy to look into it if you can give me a couple of pointers!

KoBeWi commented on editor/editor_help.h:
```diff
@@ -195,10 +197,19 @@ class EditorHelp : public VBoxContainer {
 	static String doc_version_hash;
 	static Thread worker_thread;
 
+	static bool script_docs_loaded;
```
```suggestion
	inline static bool script_docs_loaded = false;
```
You can do the same with other properties.

KoBeWi commented:
> We don't have a way to change only part of the file, so the entire disk cache will be invalidated whenever any single script is saved.

Yeah but until you modify a script, the cache is valid. Especially relevant for built-in scripts, because they currently don't have documentation generated, so they don't invalidate cache.
Also no reason to delete the file - you can just overwrite it when saving, no?

> Not sure what the folding is

Inspector folding. When you edit resource in the inspector it saves what property was folded. But I never edited the file, so I was just surprised. Likely unrelated to this PR.

KoBeWi commented on editor/editor_help.cpp:
```diff
@@ -2974,7 +2997,138 @@ void EditorHelp::_gen_extensions_docs() {
+	}
+
+	Array classes = script_doc_cache_res->get_meta("classes", Array());
+	for (const Dictionary dict : classes) {
```
```suggestion
	for (const Dictionary &dict : classes) {
```
It's interesting that you don't need to use Variant 🤔

anvilfolk replied:
I believe it's implicitly converting the Variant into a `const Dictionary`. `const Dictionary&` doesn't compile, since it needs to do the casting/conversion from `const Variant&` into`const Dictionary&`. But then the conversion happens elsewhere and what we expect to be inside `classes` is less clear!

KoBeWi commented on editor/editor_help.cpp:
```diff
@@ -2974,7 +2997,138 @@ void EditorHelp::_gen_extensions_docs() {
+	}
+
+	// Process postponed doc changes, likely added by EditorFileSystem's scans while the cache was loading in EditorHelp::worker_thread.
+	for (const DocData::ClassDoc &cd : _docs_to_add) {
```
You can move adding after deletion, so deleting is faster (less elements to lookup).

KoBeWi commented on editor/editor_help.h:
```diff
@@ -195,10 +197,19 @@ class EditorHelp : public VBoxContainer {
 	static Thread worker_thread;
 
+	inline static bool script_docs_loaded = false;
+	inline static Vector<DocData::ClassDoc> _docs_to_add;
+	inline static Vector<String> _docs_to_remove;
+	inline static Vector<String> _docs_to_remove_by_path;
```
These can be LocalVectors.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -829,6 +830,11 @@ void EditorNode::_notification(int p_what) {
 #endif
 		} break;
+
+		case NOTIFICATION_PREDELETE: {
```
This could be `EXIT_TREE` instead, but if it it works 🤷‍♂️

anvilfolk commented:
> Yeah but until you modify a script, the cache is valid.

Yes! But it was already loaded on startup, so it did its job. My claims are these:
1) It's unlikely someone will use Godot without touching any script (thus invalidating the cache), so keeping the cache is useful only in few usecases.
2) The cache saves quickly on exit, so it's not worth the extra code just for the few usecases above.
3) If performance is an issue for large projects, it will be an issue in most uses of the editor (when scripts are changed) so we need a different solution anyways, e.g. project setting to disable script docgen.

> Also no reason to delete the file - you can just overwrite it when saving, no?

We delete it so we don't load docs inconsistent with the scripts if Godot crashes and you don't get to save the disk cache on editor exit. Rewriting the entire cache any time a script is saved feels overkill and a performance issue! :)

KoBeWi commented:
Tested with my very big project™ and the editor is stuck on reopening scenes:
![image](https://github.com/user-attachments/assets/6d0664a4-68b5-492b-b3c5-391e796eed0b)
It's one of the initial steps when launching. Maybe something gets locked by the thread?

anvilfolk commented:
I also noticed I am getting `core\object\object.cpp:2342 - ObjectDB instances leaked at exit (run with --verbose for details).` when the cache is regenerated. I'll need to test with a smaller project so I can use `--verbose`.

Thanks for testing on a large project, I only tested on project that had a lot of .gd files, not scenes. I thought most of the threading issues would've been resolved but I'll take another look.

That could be where it hangs if it's regenerating the cache - but it sounds like it's a full stop. Could be any call to `EditorHelp::get_doc_data()` during the initial filesystem scan 🤔

KoBeWi commented:
Stuck trace:
```
Thread::wait_to_finish() Line 87 (c:\godot_source\core\os\thread.cpp:87)
EditorHelp::_wait_for_thread() Line 2866 (c:\godot_source\editor\editor_help.cpp:2866)
EditorHelp::get_doc_data() Line 3329 (c:\godot_source\editor\editor_help.cpp:3329)
Shader::get_shader_uniform_list(List<PropertyInfo,DefaultAllocator> * p_params, bool p_get_groups) Line 177 (c:\godot_source\scene\resources\shader.cpp:177)
ShaderMaterial::_get_property_list(List<PropertyInfo,DefaultAllocator> * p_list) Line 234 (c:\godot_source\scene\resources\material.cpp:234)
ShaderMaterial::_get_property_listv(List<PropertyInfo,DefaultAllocator> * p_list, bool p_reversed) Line 95 (c:\godot_source\scene\resources\material.h:95)
Object::get_property_list(List<PropertyInfo,DefaultAllocator> * p_list, bool p_reversed) Line 512 (c:\godot_source\core\object\object.cpp:512)
Resource::duplicate_for_local_scene(Node * p_for_scene, HashMap<Ref<Resource>,Ref<Resource>,HashMapHasherDefault,HashMapComparatorDefault<Ref<Resource>>,DefaultTypedAllocator<HashMapElement<Ref<Resource>,Ref<Resource>>>> & p_remap_cache) Line 289 (c:\godot_source\core\io\resource.cpp:289)
SceneState::make_local_resource(Variant & p_value, const SceneState::NodeData & p_node_data, HashMap<Ref<Resource>,Ref<Resource>,HashMapHasherDefault,HashMapComparatorDefault<Ref<Resource>>,DefaultTypedAllocator<HashMapElement<Ref<Resource>,Ref<Resource>>>> & p_resources_local_to_sub_scene, Node * p_node, const StringName p_sname, HashMap<Ref<Resource>,Ref<Resource>,HashMapHasherDefault,HashMapComparatorDefault<Ref<Resource>>,DefaultTypedAllocator<HashMapElement<Ref<Resource>,Ref<Resource>>>> & p_resources_local_to_scene, int p_i, Node * * p_ret_nodes, SceneState::GenEditState p_edit_state) Line 649 (c:\godot_source\scene\resources\packed_scene.cpp:649)
SceneState::instantiate(SceneState::GenEditState p_edit_state) Line 369 (c:\godot_source\scene\resources\packed_scene.cpp:369)
PackedScene::instantiate(PackedScene::GenEditState p_edit_state) Line 2171 (c:\godot_source\scene\resources\packed_scene.cpp:2171)
SceneState::instantiate(SceneState::GenEditState p_edit_state) Line 232 (c:\godot_source\scene\resources\packed_scene.cpp:232)
PackedScene::instantiate(PackedScene::GenEditState p_edit_state) Line 2171 (c:\godot_source\scene\resources\packed_scene.cpp:2171)
EditorNode::load_scene(const String & p_scene, bool p_ignore_broken_deps, bool p_set_inherited, bool p_clear_errors, bool p_force_open_imported, bool p_silent_change_tab) Line 4045 (c:\godot_source\editor\editor_node.cpp:4045)
EditorNode::_load_open_scenes_from_config(Ref<ConfigFile> p_layout) Line 5370 (c:\godot_source\editor\editor_node.cpp:5370)
EditorNode::_load_editor_layout() Line 5270 (c:\godot_source\editor\editor_node.cpp:5270)
```
idk if it ever stops, I gave up after 8 mins.

anvilfolk commented:
Thank you for digging into it so much, that's really really helpful!

It sounds like generating the cache for the first time causes the freeze. It may or may not be related to there being shaders, or to the fact that shader documentation may be considered "script docs". I'll look into it, and address the other stuff too!

anvilfolk commented:
I've been wracking my brain on how this race condition could happen but I'm not getting anywhere. @KoBeWi were you perhaps running a `-use_asan=true use_ubsan=true` build or with `--verbose`? Both of those things have, in my experience, very significantly increased the time it takes to regenerate the cache. But still nowhere near 8 minutes 🤔 If it hangs consistently, would you be able to run it with `--verbose`? The doc cache regen should attempt to load every single script.

I could see ``EditorHelp::get_doc_data()` locking only if `ResourceLoader::load(p_dir->get_file_path(i));`, which is called from `EditorHelp::worker_thread` in `EditorHelp::_reload_scripts_documentation()` hangs from interacting with some other thread also loading stuff. Could that be a possibility?

KoBeWi commented:
I'm using `dev_build=yes scu_build=yes fast_unsafe=yes`. The last file printed by verbose is an image, preceded by a scene.

I can send you the project if you can't figure it out.

anvilfolk commented:
That project would help - I could at least inspect where the threads are and figure out what's causing the deadlock.

Turns out shaders are considered scripts for doc purposes and will have their documentation cached when they are compiled, e.g. when a scene with a shader is loaded in the editor. Might need to look into how to remove those docs if the shader is ever deleted.

huwpascoe commented on editor/editor_help.h:
```diff
@@ -195,10 +197,19 @@ class EditorHelp : public VBoxContainer {
 	static String doc_version_hash;
 	static Thread worker_thread;
 
+	inline static bool script_docs_loaded = false;
```
These are static inline. That means Editor help will get unique instances of these variables per cpp file that includes them. Is this intentional?

KoBeWi replied:
Isn't `static inline` just a way to define static variable in a header, without having to declare it in header and define in cpp file? Or are they not equivalent?

huwpascoe replied:
They're not equivalent. It's like inline functions. The cpp file (translation unit) reads it as "I get my own copy to optimize what I like"

KoBeWi replied:
That's unfortunate, because they are used as equivalent in lots of places 💀

huwpascoe replied:
Well. Seems static inline is a new c++17 feature. How ignorant I am!

anvilfolk commented:
Latest update:
1) Changed `_script_docs_loaded` to `SafeFlag`
2) Removed race condition in reading/writing to `docs_to_add/remove/etc` by pulling setting `_script_docs_loaded` to before those are read. In that case, other threads will wait until this thread is finished to add those docs.


I noticed that `filesystem_changed` or `sources_changed` aren't always emitted. @Hilderin, I'm thinking it may be safer to emit a new `first_scan_complete` signal to make sure. What are your thoughts?

anvilfolk commented:
New update:
- Fixed reason for deadlock found by @KoBeWi, thanks for sharing the project so I could replicate it
- Changed `filesystem_changed` which isn't always fired, to `sources_changed` which should always be fired.

The reason for the deadlock was that non-main threads like the EditorHelp worker thread will queue resource load tasks and wait for them to be dispatched (I assume) on the main thread and loaded from Godot's general worker threads. But, if the main thread waits on the EditorHelp worker thread by calling e.g. `EditorHelp::get_doc_data()` before dispatching the load tasks, then it deadlocks.

I added a verbose error message to catch when deadlocks might happen.

Also, force-reloading all the scripts from a worker thread in a larger project like @KoBeWi's has resulted in `Invalid task ID.` a few times @RandomShaper.

anvilfolk commented:
I'm going to create a new worker thread just for the `ResourceLoader::load()` calls. It should prevent all deadlocks without having to change all existing calls to `EditorHelp::get_doc_data()` to use the new helper functions.

anvilfolk commented:
Latest update:
- Added a `loader_thread` to `EditorHelp`. It is used to make the `ResourceLoader::load()` calls, which were deadlocking when the main thread wait for the `worker_thread`. There should be no more deadlocking possible at all.
- Changed EditorHelpBit and ScriptEditorPlugin to use new EditorHelp doc access methods. Can be replicated throughout the rest of the codebase where appropriate, but it's not necessary.
- Will run more comprehensive tests when I get back to the right computer

magian1127 commented:
> EDIT2: I saved the cache as .tres and inspected it. It seems to also contain shaders?
> 
> It shows in search, but you can't open it !

@KoBeWi 

This was caused by my previous PR, and I've now submitted a fix. #97616

anvilfolk commented:
Testing is showing docs are not when changed when Godot is closed, or when changed from outside Godot with it open. Looking into it. Could be either because something changed in EditorFileSystem which stopped `_update_scripts_documentation()` from being called, or from my changing from `filesystem_changed` to `sources_changed`.

anvilfolk commented on editor/editor_help.cpp:
```diff
@@ -2973,8 +2995,166 @@ void EditorHelp::_gen_extensions_docs() {
+void EditorHelp::regenerate_script_doc_cache() {
+	if (EditorFileSystem::get_singleton()->is_scanning()) {
+		// Wait until EditorFileSystem scanning is complete to use updated filesystem structure.
+		EditorFileSystem::get_singleton()->connect(SNAME("sources_changed"), callable_mp_static(_regenerate_script_doc_cache), CONNECT_ONE_SHOT);
```
Changed to `sources_changed` from `filesystem_changed`. It looks like that might be a more reliable signal, since in some situations where no change has happened, `filesystem_changed` may not emit. I also like that `OS::get_singleton()->benchmark_end_measure("Editor", "First Scan");` happens in `EditorNode::_sources_changed`, which is good evidence that it's reliable to detect the end of the first scan.

anvilfolk commented on editor/editor_help.cpp:
```diff
@@ -2973,8 +2995,166 @@ void EditorHelp::_gen_extensions_docs() {
+	ERR_FAIL_COND_MSG(!ProjectSettings::get_singleton()->is_project_loaded(), "Error: cannot load script doc cache without a project.");
+	ERR_FAIL_COND_MSG(!ResourceLoader::exists(get_script_doc_cache_full_path()), "Error: cannot load script doc cache from inexistent file.");
+
+	Ref<Resource> script_doc_cache_res = ResourceLoader::load(get_script_doc_cache_full_path(), "", ResourceFormatLoader::CACHE_MODE_IGNORE);
```
Added `CACHE_MODE_IGNORE` here since it feels silly to cache something we're deleting anyway :)

Perhaps this is worth doing for the native docs as well?

anvilfolk commented:
- Issue mentioned above fixed in #97710
- See comments above for latest changes: changed signal used and added `CACHE_MODE_IGNORE` when loading script doc cache.

Ran the following tests.

With Godot open:

| | Add script | Modify script | Remove script |
| ---: | :---:  | :---: | :---: |
| Internal script editor |✅ |✅ |✅ |
| External script editor |✅ |✅ |✅ |

With Godot closed:
|  | Regenerate docs| Added script | Modified script | Removed script |
| ---:         |     :---:      |      :---:      |     :---:      |     :---:      |
| Empty project, no extra scripts   | ✅ |✅ |✅ |✅ |
| Empty project, extra scripts   | ✅ |✅ |✅ |✅ |
| Complex project, no extra scripts   | ✅ ⚠️  |✅ |✅ |✅ |
| Complex project, extra scripts   | ✅⚠️  |✅ |✅ |✅ |

The extra scripts are 4096 extra scripts that exist just to make script regen take longer and hopefully bring concurrency issues to light. None did.

Notes:
- The ⚠️ warnings ⚠️ above refer to getting the occasional `Invalid task ID` error, presumably because `loader_thread` is loading gdscript files that may be used & being loaded elsewhere. It is more of a concurrent resource loading problem than anything else.
- I also ran into an issue where Godot would straight up close repeatedly with the complex project, but never within VS. It stopped happening, and I couldn't get a log or replicate at all after a while.
- I'm also getting `ObjectDB` leaked at exit, but I don't really see how that might be from this. Unless it's something to do with saving a resource on exit? I'm not sure.

RandomShaper commented:
I've been trying to diagnose the two issues related to `ResourceLoader`:
1. Deadlock if not using a thread to start the load.
2. _Invalid task id_ error.

But I don't know how to reproduce them. I know no. 1 has been worked around, but I'd still like to see the deadlock happening in front of me with the version of the code prior to the workaround that I can retrieve from GitHub.

Also, regarding no. 1, could you tell me what happens if instead of using the blocking `ResourceLoader::load()`, you use `ResourceLoader::load_threaded_request()` with `p_use_sub_threads = true`?

anvilfolk commented:
Thank you for following up! <3 I'll try to find some time in the coming days to get back to a previous commit and try your suggestion! :)

If @KoBeWi feels good about sharing the project with you, you should be able to replicate 2. by opening it when the script docs cache isn't present (because it was never generated it or you deleted it).

Jowan-Spooner commented:
Any update on this? Would be great to see this in 4.4!

anvilfolk commented:
Thanks for the bump!

Been stuck at home with my toddler for over 6 weeks because of a COVID infection, possibly with some false positive tests (careful with the QuickVue brand 😅 ). I'll prioritize looking into @RandomShaper's suggestions once 2025 hits and life hopefully normalizes!

Some thoughts beforehand though!

1. The deadlock was just a feature of the main thread waiting on the worker thread, but the worker thread using `ResourceLoader::load()` which depends on the main thread to dispatch tasks. The addition of a loading thread the main thread does _not_ wait on fixes this.
2. Using `load_threaded_request()` means following up `load_threaded_get_status()` for hundreds/thousands of files so that we know when we're ready to copy data over for usage in `EditorHelp`. There does not appear to be a callback for when a resource finishes loading. It might be cleaner to just use two threads as-is?

@KoBeWi @RandomShaper did you evershare the project so the Invalid Task ID could be investigated? I believe this PR exposes the issue but does not necessarily cause it, so perhaps that shouldn't be seen as a blocker here?

AThousandShips reviewed (approved):
Haven't tested but code looks good

KoBeWi commented:
Needs another rebase.

anvilfolk commented:
Only did some basic tests but don't think the rebase broke anything.

dalexeev reviewed (approved):
I am not familiar with this area, but overall the code looks good to me and I believe that anvilfolk and Hilderin have done a good job!

I have tested these changes on several small-medium projects and have not noticed any major issues. The only thing is that the editor startup time without cache has increased noticeably. But this is quite expected, probably caused by the fact that now all project scripts are compiled. Restarting the editor takes much less time.

We could probably avoid full analysis and compilation of all scripts, for full project documentation only the interface resolution is required, function body analysis and script compilation are not really needed. But this is problematic to do, since `GDScriptCache` is a rather complicated system and we are forced to compile all scripts to get full documentation.

I also noticed a phantom analyzer error with a circular dependency. But it immediately disappeared when I opened the dependent script and I'm not sure if it's related to this PR, maybe it's caused by the missing editor cache.

So personally I would highly support merging this PR for 4.4. It would be a significant QoL improvement and would fit well with the new documentation tooltips feature. But of course it has concerns about edge cases and editor performance regressions for large projects.

It would probably be better to merge this early into 4.5 dev for more testing. But I think it also makes sense to test it in 4.4 beta. In the worst case, we can revert it before release and postpone to 4.5 if we find major issues.

KoBeWi commented:
Tested with a [complex project](https://github.com/KoBeWi/Metroidvania-System). The docs weren't loaded during initial launch, I had to open and re-save the scripts. This also caused an error print (something about missing doc data, I didn't copy it). Afterwards the docs work correctly, including between editor sessions.

This only happens if the project was already open and imported before. Opening without `.godot` folder will properly initialize the cache, so it's probably not worth fixing. I did not observe other problems.

anvilfolk commented on editor/doc_tools.cpp:
```diff
@@ -368,6 +368,15 @@ void DocTools::remove_doc(const String &p_class_name) {
+	for (KeyValue<String, DocData::ClassDoc> &E : class_list) {
+		if (E.value.is_script_doc && E.value.script_path == p_path) {
+			remove_doc(E.key);
+			return;
```
Thanks for looking into it!

> We could probably avoid full analysis and compilation of all scripts.
Yes! There are a lot of places where we are using `ResourceLoader::load(some_script_file)`, which full compiles something, in order to get documentation. This would be a wonderful thing to optimize in the future, but there's definitely a ton of technical challenges to overcome there!

The circular dependency is a little worrying for sure. I'm at a loss as to how it may be happening though, since the loading thread just loads scripts one by one, so circular dependencies should be handled as they normally are 🤔

anvilfolk replied:
> Can multiple `ClassDoc`s have the same `script_path`? For example inner classes.

I'll need to double-check, good catch! I believe fqcn's have a separator from the path for inner classes but I'm not sure what their `script_path` is.

akien-mga commented:
We're reaching feature freeze for 4.4 now, so I'll punt this to 4.5, but it should be ready to merge early on for 4.5-dev1.

This could have technically made it in 4.4 but since it was approved a bit late in the cycle, I'm not confident we have enough time to really battle test this before stable, especially on very big projects where I'm concerned about some of the performance implications.

On the flip side, folks will have another reason to wait for Godot mere days after 4.4-stable is released :P

anvilfolk commented:
Toddler is on/off sick so it's been hard to find time to follow up on issue reports here, so I think that's a good call too.

If I don't get to figure out what's happening, we can always iterate in 4.5 dev, and work on project options, etc.

anvilfolk commented:
Oops, misclick. I rebased locally a while back, will try to get this done soon! <3

anvilfolk commented:
Rebased, did some basic tests that all looked OK!

Some quick reminders:
- Project load times w/ non-existent disk cache are increased significantly for large projects (since all scripts are force-compiled), but only slightly when disk cache exists.
- Might expose some issues with threaded resource loading resulting in `Invalid Task ID` - will need more reports when this goes out. Possibly related to tool scripts and/or circular dependencies.

In the future we should add options to:
- Disable script doc cache in situations where a project / Godot is unstable so folks don't have to force-recompile all scripts all the time.
- Force regenerate the script cache from the editor in case it becomes inconsistent so folks don't need to go delete the file from `.godot` manually.

Repiteo commented:
Thanks!
